### PR TITLE
Check if rnn weights need to be flattened

### PR DESCRIPTION
**Pull request by jithunnair-amd**
cuDNN needs it, MIOpen doesn't. However, since it seems to be the PyTorch preference to not introduce ROCm-specific logic in the python layer, we need to add a C++ function to detect if rnn weight flattening is needed.

This PR will be needed to fix the rnn unit test errors arising for PR #33837.

### CODE REVIEW

**Comment by jithunnair-amd** [comment]
cc @iotamudelta @ashishfarmer @t-vi

**Comment by jithunnair-amd** [comment]
@ezyang Can you please review these changes? 
In commit [f382536](https://github.com/pytorch/pytorch/pull/34265/commits/f382536c7d44fb4208525b098ec3d32c6a30aeaa), I introduced the `_use_cudnn_rnn_flatten_weight` function specifically for CUDA dispatch, but had to add a dummy parameter since I'd get runtime errors about the function needing at least one argument of type `Tensor` or `TensorList`. In commit [136c197](https://github.com/pytorch/pytorch/pull/34265/commits/136c197794f38fc7813498da2c31cc5b178883ea), I refactored the function to not have any dummy parameters, but that doesn't work for CPU-only builds.
Can you provide some guidance on what would be the best way to achieve the objective here?

**Comment by jithunnair-amd** [comment]
Moved the `_use_cudnn_rnn_flatten_weight` function from `aten/src/ATen/native/cudnn/RNN.cpp` to `aten/src/ATen/native/RNN.cpp` in [81537a1](https://github.com/pytorch/pytorch/pull/34265/commits/81537a149332733db01683ec43b10638fd3610e2) so that the CPU builds pass. H/T: @t-vi 

I think this PR is ready for review now. The failing builds are due to a "docker image not found" error.

**Review by iotamudelta — APPROVED** [review]
LGTM. We've discussed this internally.

**Review by t-vi — APPROVED** [review]
Seems good to me. Thanks!

**Comment by ezyang** [comment]
I'd appreciate it if it were more obvious that this was an NVIDIA versus AMD thing via  comment ;)

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -856,6 +856,10 @@ std::tuple<io_type, Tensor, Tensor> _lstm_impl(
 
 } // anonymous namespace
 
+bool _use_cudnn_rnn_flatten_weight() {
```
Comment (jithunnair-amd) [inline aten/src/ATen/native/RNN.cpp]: @ezyang Would this be an appropriate point for the comment that mentions that it's a cuDNN vs MIOpen thing?

Comment (ezyang) [reply aten/src/ATen/native/RNN.cpp]: yep

Comment (jithunnair-amd) [reply aten/src/ATen/native/RNN.cpp]: I'm sorry, I was about to put in the comment but it looks like it got merged already. :) There's always `git blame` which will lead someone to this PR :)

**Pull request closed without merging**